common/materialContainer/MaterialContainer.h

```diff
@@ -13,58 +13,79 @@
 #include <stdlib.h>
 #include <cstdlib>
 #include <type_traits>
+#include <map>
+#include <memory>
+#include "PerlinNoise.h"
+#include "block.h"
+
+
+
+extern const double GROUND_OFFSET;
+extern const double GROUND_SMOOTHNESS;
+extern const double BIOME_OFFSET;
+extern const double BIOME_SMOOTHNESS;
+extern const double BIOSPHERE_OFFSET;
+extern const double BIOSPHERE_SMOOTHNESS;
+
 
 class Material
 {
 private:
   std::string name;
+  BlockType type;
   double freq;
-
+  static std::map <std::string, std::shared_ptr<Material>> materialMap;
 public:
-  Material(std::string name, double freq) : name(name), freq(freq) {}
+  Material(std::string inName, BlockType inType, double inFreq) : name(inName), type(inType), freq(inFreq) {}
+  static std::shared_ptr<Material> makenew(std::string name, BlockType type, double freq);
   double getFreq() { return freq; }
   std::string getName() { return name; }
-  static Material Coal();
-  static Material Iron();
-  static Material Soil();
-  static Material Gravel();
-  static Material Sand();
-  static Material Diamond();
-  static Material Stone();
-  static Material Gold();
-  Material getMaterial();
+  std::shared_ptr<Material> getMaterial(int x, int y, int z);
+  BlockType getType(){ return type; }
+  void setType(BlockType BT){ type = BT; }
+  static std::shared_ptr<Material> getFromType(BlockType BT);
+  static std::shared_ptr<Material> Coal();
+  static std::shared_ptr<Material> Iron();
+  static std::shared_ptr<Material> Soil();
+  static std::shared_ptr<Material> Gravel();
+  static std::shared_ptr<Material> Sand();
+  static std::shared_ptr<Material> Diamond();
+  static std::shared_ptr<Material> Stone();
+  static std::shared_ptr<Material> Gold();
+  static std::shared_ptr<Material> Default();
 };
 
 template <class T> class MaterialContainer
 {
-protected:
+ protected:
+  double translate(int v){return((v+offset)* smoothness);}
+  static PerlinNoise noise;
   // temp PerlinNoise until intergration with project PerlinNoise
-  double PerlinNoise()
+  double PN(int x, int y, int z)
   {
-    return (static_cast<double>(rand()) / static_cast<double>(RAND_MAX));
+    static const int DEPTH=5;
+    return noise.turbulence3D(translate(x), translate(y), translate(z), DEPTH);
+    //return (static_cast<double>(rand()) / static_cast<double>(RAND_MAX));
   }
```

Comment:
While I agree that one should not reinvent the wheel I do think there is something to be said for using the code that was written for the sake of this class.

But thank you for the insight, I was unaware that GLM had perlin noise.